cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -915,6 +918,57 @@ public long getProgramRunCount(ProgramId programId) throws NotFoundException {
     });
   }
 
+  @Override
+  public List<ProgramHistory> getRuns(Collection<ProgramId> programs, ProgramRunStatus status, long startTime,
+                                      long endTime, int limit, Predicate<RunRecordMeta> filter) {
+    return Transactionals.execute(transactional, context -> {
+      List<ProgramHistory> result = new ArrayList<>(programs.size());
+      AppMetadataStore appMetadataStore = getAppMetadataStore(context);
+
+      for (ProgramId program : programs) {
+        ApplicationMeta appMeta = appMetadataStore.getApplication(program.getParent());
```

Comment:
You can use `getApplicationsForAppIds` in `AppMetadatStore` to batch get the app meta. It will return a map of app id to app meta.

Reply:
done